05-mvc-with-rails/error-handling/readme.md

```diff
@@ -3,14 +3,461 @@
+```
+
+That `ROLLBACK` indicates that ActiveRecord tried running a SQL command, but it was unsuccessful, so it's undoing -- or "rolling-back" -- any changes that it made.
+
+### With a bang
+
+Now try the same command, but put a bang `!` at the end of `.create`:
+
+```
+$ billy = Artist.create!(name: "Billy Ray Cyrus")
+```
+
+#### What's the difference between `.create` and `.create!`?
+
+Adding in a bang makes the app throw a fatal error -- that is, "break" -- if a validation fails. Without the bang, it fails "silently".
+
+You can add a bang to both `.create` and `.save`. 
+
+#### It's considered good practice to always use bangs on these methods. Why?
+
+# Error Handling
+
+## Framing
+
+The discussion on validations has so far shown you about 18 different ways a user can "break" your app. It's useful for *us* when the app breaks because we can see where all the errors are... but we don't want our users to know!
+
+If only there was something that could *rescue* our app whenever it's about to break!
+
+## Begin and Rescue
+
+Let's intentionally break something the app.
+
+Go to the Artist model, and replace it with the following:
+
+```rb
+class Artist < ActiveRecord::Base
+  has_many :songs
+  validates :name, presence: true
+end
+```
+
+Now go to the Artists controller. Change the `create` action so that it renders the Artists#new view.
+
+```rb
+  def create
+    @artist = Artist.create!(artist_params)
+    render :new
+  end
+```
+
+Finally, go to `localhost:3000/artists/new`, leave the form blank, and submit it. You should get an error saying "Validation failed: Name can't be blank".
+
+Now change the `create` action to look like the following:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue
+    end
+    render :new
+  end
+```
+
+Submit the form again. It should render the Artists#new page no problem.
+
+The `begin` tells Ruby, "Start listening for errors." The `rescue` tells Ruby to "rescue" the app from breaking when any errors come up.
+
+**Every `begin` must have a `rescue`, and vice-versa. It must also have an `end`.**
+
+### Error messages
+
+That's all very well, but it's not very useful.
+
+#### Why is this error-suppresser not very useful?
+
+If someone's interacting with your app and does something wrong, they want to know what they did wrong. Silent errors are just as bad as app-breaking errors!
+
+Let's make an error message that actually shows up somewhere:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue
+      @error = "You done broke it."
+    end
+    render :new
+  end
+```
+
+We can make errors show up in the application layout, `views/layouts/application.html.erb`:
+
+```erb
+<body>
+  <h1>Tun.r</h1>
+  <p><%= @error %></p>
+```
+
+That at least tells us something went wrong, but doesn't tell us *what*.
+
+### StandardError
```

Comment:
This is an awfully long trip down a path of syntax we would never recommend, just to show some examples of rescue.  Working on PR for using standard rails `if save` syntax.  Recommend no further work on this area.